relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -1,170 +1,219 @@
 //! Functionality for calculating if a trace should be processed or dropped.
 //!
-use chrono::{DateTime, Utc};
 use std::net::IpAddr;
 
+use chrono::{DateTime, Utc};
+
 use relay_common::{ProjectKey, Uuid};
 use relay_general::protocol::Event;
-use relay_sampling::{DynamicSamplingContext, RuleId, SamplingConfig, SamplingMode};
+use relay_sampling::{
+    DynamicSamplingContext, MatchedRuleIds, RuleType, SamplingConfig, SamplingMode,
+};
 
 use crate::actors::project::ProjectState;
 use crate::envelope::{Envelope, ItemType};
 
-macro_rules! or_ok_none {
+macro_rules! no_match_if_none {
     ($e:expr) => {
         match $e {
             Some(x) => x,
-            None => return Ok(None),
+            None => return SamplingMatchResult::NoMatch,
         }
     };
 }
 
 /// The result of a sampling operation.
-#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub enum SamplingResult {
     /// Keep the event. Relay either applied a sampling rule or was unable to parse all rules (so
     /// it bailed out)
     Keep,
     /// Drop the event, due to the rule with provided identifier.
-    Drop(RuleId),
+    Drop(MatchedRuleIds),
+}
+
+/// The result of the dynamic sampling matching executed on the sampling config.
+#[derive(Clone, Debug, PartialEq)]
+enum SamplingMatchResult {
+    /// The incoming event/dynamic sampling context matched the sampling configuration.
+    Match {
+        sample_rate: f64,
+        matched_rule_ids: MatchedRuleIds,
+        seed: Uuid,
+    },
```

Comment:
This looks very similar to the `SamplingMatch` defined in `relay-sampling`. Can we just use an `Option<SamplingMatch>` instead of this enum?

Reply:
They look similar but we actually change the sample rate, I would prefer to keep this `Match` and `NoMatch` for expressiveness and I can use the `SamplingMatch` inside the `Match` variant of the enum.